engine-precompiles/src/modexp.rs

```diff
@@ -56,13 +56,15 @@ impl<HF: HardFork> ModExp<HF> {
         let mod_start = exp_end;
 
         let base = parse_bytes(input, base_start, base_len, BigUint::from_bytes_be);
```

Comment:
I probably don't fully understand the vulnerability but can the same issue occur if only the base is provided but not the exponent and modulus? Should parsing the base be moved into the else branch as well?

Reply:
The scope of this PR is to solve the out-of-memory issue in case of passing a very big exponent. For the base and modulus are not the case as it will revert with `out-of-gas` in the `required_gas()` before executing `runn_inner()`

Reply:
Thanks for you vigilance on this @mandreyel ! I think it is ok to eagerly parse the `base` value because it is accounted for in the gas cost of the precompile. I'll try to explain this in more detail below.

First of all, the basic principle public blockchains that have smart contracts operate on is the following: users can do arbitrary computations (thus consuming arbitrary resources on the nodes of the blockchain network) so long as the user pays for them. The metric used to measure how much the user has to pay is called "gas". The idea is that operations which would crash a node (e.g. allocating terabytes of memory) are prohibitively expensive in terms of gas and therefore no user can cause those operations to happen because they will lack the funds to do it. [As an aside, our particular project is extra complicated in this regard because we are running the EVM, which has its own gas system, inside the Near runtime, which also has a gas system; so we need to make sure that operations are adequately charged in both systems, but I won't go into details on that complexity here]

Given this background, take a look at the [required_gas function](https://github.com/aurora-is-near/aurora-engine/blob/develop/engine-precompiles/src/modexp.rs#L149) for this precompile. The key point is that it has a special case for when the base and modulus are both zero, independent of the value of the exponent. So input with no base or modulus is very cheap in terms of gas, but our implementation did not have the same short-circuit, thus allowing malicious users to perform operations they were not adequately charged for. This PR fixes the issue by only parsing the exponent if it is needed for a calculation.

To directly address your question about the base, notice that the [gas cost for the base is quadratic in its size](https://github.com/aurora-is-near/aurora-engine/blob/develop/engine-precompiles/src/modexp.rs#L141-L146). Therefore, users can input very large base values, but they will also be charged very large amounts or else such input will be rejected.

Reply:
That makes sense, thank you for the in-depth explanation @birchmd. Resolving.